finality-aleph/src/testing/proposal.rs

```diff
@@ -0,0 +1,270 @@
+            false,
+        )
+        .await;
+
+    let incorrect_branch = vec![
+        blocks[0].clone(),
+        blocks[1].clone(),
+        blocks[3].clone(),
+        blocks[5].clone(),
+    ];
+    let proposal = proposal_from_blocks(incorrect_branch);
+    verify_proposal_status(
+        &mut cached_cip,
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[1].header.hash());
+    verify_proposal_status(
+        &mut cached_cip,
+        &mut aux_cip,
+        &proposal,
+        Pending(TopBlockImportedButIncorrectBranch),
+    );
+
+    chain_builder.finalize_block(&blocks[10].header.hash());
+    verify_proposal_status(&mut cached_cip, &mut aux_cip, &proposal, Ignore);
+}
+
+#[tokio::test]
+async fn pending_top_block_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_branch_upon(&chain_builder.genesis_hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+    chain_builder.import_branch(blocks.clone(), false).await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..len].to_vec();
+        let proposal = proposal_from_blocks(blocks_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Finalize(proposal.top_block()),
+        );
+    }
+}
+
+#[tokio::test]
+async fn hopeless_forks_handled_correctly() {
+    let (mut chain_builder, mut cached_cip, mut aux_cip) = prepare_proposal_test();
+
+    let blocks = chain_builder
+        .build_and_import_branch_upon(
+            &chain_builder.genesis_hash(),
+            MAX_DATA_BRANCH_LEN * 10,
+            false,
+        )
+        .await;
+
+    let forks = chain_builder
+        .build_branch_upon(&blocks[2].header.hash(), MAX_DATA_BRANCH_LEN * 10)
+        .await;
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
+        );
+    }
+
+    chain_builder.finalize_block(&blocks[2].header.hash());
+
+    for len in 1..=MAX_DATA_BRANCH_LEN {
+        let fork_branch = forks[0..len].to_vec();
+        let proposal = proposal_from_blocks(fork_branch);
+        verify_proposal_status(
+            &mut cached_cip,
+            &mut aux_cip,
+            &proposal,
+            Pending(PendingTopBlock),
```

Comment:
I'm wondering whether there is any difference between such hopeless fork and the imported one :thinking:

Reply:
I'm not entirely sure what you are referring to -- are you wondering what would happen if `fork` was imported before this line? In that case the proposal would be "finalizable".

Reply:
actually, I was thinking about `TopBlockImportedButNotFinalizedAncestor` status (which I guess would be correct at this moment), but you are already checking this case in the other testcase, so resolving the discussion